File: app/src/main/java/com/bot/man/api/ApiHandler.java
package com.bot.man.api;

import android.util.Log;

import com.bot.man.model.data.Response;
import com.bot.man.viewmodel.MainViewModel;

import org.jetbrains.annotations.NotNull;

import retrofit2.Call;
import retrofit2.Callback;

public class ApiHandler {
	private static ApiHandler INSTANCE;

	private ApiHandler() {
	}

	public static ApiHandler getInstance() {
		if (INSTANCE == null) {
			INSTANCE = new ApiHandler();
		}
		return INSTANCE;
	}

	public void fetchData(
			MainViewModel.OnResultListener mOnResultListener, String API_KEY) {
		new HandlerThread(mOnResultListener, API_KEY).start();
	}

	private static class HandlerThread extends Thread {
		private static final String TAG = "BotHandlerThread";
		private String API_KEY;
		private MainViewModel.OnResultListener mOnResultListener;

		HandlerThread(
				MainViewModel.OnResultListener onResultListener,
				String API_KEY) {
			mOnResultListener = onResultListener;
			this.API_KEY = API_KEY;
		}

		@Override
		public void run() {
			//TODO: Fetch results and call the onResultListener
			// .onUpdateChanged(Response response).

			Helper.getInstance(API_KEY).fetchGetUpdates()
			      .enqueue(new Callback<Response>() {
				      @Override
				      public void onResponse(
						      @NotNull Call<Response> call,
						      @NotNull retrofit2.Response<Response> response) {
					      if (response.body() != null) {
						      Log.e(TAG,
						            "onResponse " + response.body().toString());
						      mOnResultListener
								      .onUpdateChanged(response.body());
					      }
					      interrupt();
				      }

				      @Override
				      public void onFailure(
						      @NotNull Call<Response> call,
						      @NotNull Throwable t) {
					      Log.e(TAG, "onFailure " + t.getLocalizedMessage());
					      mOnResultListener.onUpdateError(t);
				      }
			      });
		}
	}
}

File: app/src/main/java/com/bot/man/api/ApiHandler.kt
package com.bot.man.api

import android.util.Log
import com.bot.man.model.data.Response
import com.bot.man.viewmodel.MainViewModel.OnResultListener
import retrofit2.Call
import retrofit2.Callback

class ApiHandler private constructor() {
    fun fetchData(
            mOnResultListener: OnResultListener, API_KEY: String) {
        HandlerThread(mOnResultListener, API_KEY).start()
    }

    private class HandlerThread internal constructor(
            private val mOnResultListener: OnResultListener,
            private val API_KEY: String) : Thread() {
        override fun run() {
            Helper.getInstance(API_KEY)?.fetchGetUpdates()
                    ?.enqueue(object : Callback<Response?> {
                        override fun onResponse(
                                call: Call<Response?>,
                                response: retrofit2.Response<Response?>) {
                            if (response.body() != null) {
                                Log.e(
                                        TAG,
                                        "Response status: " +
                                                response.body()!!.ok)
                                mOnResultListener
                                        .onUpdateChanged(response.body())
                            }
                            interrupt()
                        }

                        override fun onFailure(
                                call: Call<Response?>,
                                t: Throwable) {
                            Log.e(
                                    TAG,
                                    "onFailure " + t.localizedMessage)
                            mOnResultListener.onUpdateError(t)
                        }
                    })
        }

        companion object {
            private const val TAG = "BotHandlerThread"
        }

    }

    companion object {
        private var INSTANCE: ApiHandler? = null

        @JvmStatic
        val instance: ApiHandler?
            get() {
                if (INSTANCE == null) {
                    INSTANCE = ApiHandler()
                }
                return INSTANCE
            }
    }
}